### DISCUSSION TITLE
Intelligent calculation of route sections

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#142] [Q&A] [answered]
**Intelligent route calculation** of individual sections instead of the whole route.

- Calculate only the changed part of the route, useful for long routes

- Use different routing options on each route section
(change routing options and select "**Recalculate section**")

- https://github.com/devemux86/cruiser/discussions/446

- Many routing actions are now done offline without calculation:
straight line, waypoint types, delete start / end, split route, etc.

- Select and combine alternatives independently in route sections

---

**Intelligent rerouting** is already available in **navigation** where we merge:
- Rerouting to the next waypoint, e.g. BRouter (offline)
- Existing route remains **unchanged**, e.g. curvy route

<img src="https://github.com/user-attachments/assets/2fc678a7-8dbd-49bb-8d93-147ef05ac0ae" width="50%"/><img src="https://github.com/user-attachments/assets/b61864ee-9553-4280-bd0a-2ef5d98abb0c" width="50%"/>

**Comment by devemux86** [comment]
When planning short trips "there and back", I set several via-points to visit and let BRouter calculate alternatives, from which I choose the most interesting one. This works great, but we don't have a choice of the course of individual sections between points, but only the entire route. 

Is it possible to change the behavior of BRouter so that individual sections are as if individual routes, with the possibility of choosing alternatives? Such a possibility would make it much easier, for example, to choose the best of the alternative return routes, along previously unvisited roads. 

This could be a precursor to the future function of selecting route parameters for individual sections 😉

_Originally posted by @Ciutnik in https://github.com/devemux86/cruiser/discussions/234#discussion-5368670._

**Comment by Ciutnik** [comment]
My problem seems to fit into this topic.

When I'm planning the route from the START point to the END point, I choose one of the alternatives that Cruiser navigates. This works very well. 

If I start the route not exactly from the above specified START point, Cruiser can report off the route and propose a recalculation (automatic recalculation is always disabled). This is also expected behavior.

But here is the problem: if a recalculation occurs, the alternative I originally chose is replaced by another route, the default first one, chosen by Cruiser (BRouter). Even if I choose to recalculate to the original START point. I can deal with this by pausing navigation to display all alternatives and selecting the one I am interested in again. But if the recalculation occurs somewhere in the middle of the route, will my chosen alternative be "forgotten"? If possible, it would be good if she was remembered.

**Reply by devemux86** [reply]
@Ciutnik it is a different issue, let's discuss it here:
- https://github.com/devemux86/cruiser/discussions/312

**Comment by Leeb64** [comment]
I actually just wanted to dig this absolutely sensible thing out of obscurity...

**Reply by devemux86** [reply]
@Leeb64 it just requires a lot of work compared to other features. :slightly_smiling_face:

**Comment by devemux86** [comment]
**Intelligent route calculation** of individual route sections is available in:
- https://github.com/devemux86/cruiser/discussions/451

Please test it well to see if everything works properly. Instructions in the announcement above: https://github.com/devemux86/cruiser/discussions/142

If you see something strange, please report the route and detailed steps so we can reproduce it.

**Reply by bios4** [reply]
Very nice feature like the manual variant - works flawlessly with a (very short) test route from "my places". 👍

> Use different routing options on each route section

Is there a plan to offer additional routing options in the future?
Currently it is limited to "normal" and "beeline"...

**Reply by WolfgangZ42** [reply]
I found an issue - hopefully you can reproduce it ...
Switching to manual route calculation inserting a waypoint (regardless if shaping or via point) takes the number of the end waypoint. Intelligent route calculation works fine. Here the screenshots:
<img height="200" src="https://github.com/devemux86/cruiser/assets/121806238/efe5a7f0-b6de-445a-8432-6cb2a42843b3"/>  <img height="200" src="https://github.com/devemux86/cruiser/assets/121806238/24b44046-e68e-42a1-9155-a5b4a1ca5d82"/>  <img height="200" src="https://github.com/devemux86/cruiser/assets/121806238/051de409-b8d6-43f2-937b-248db8ab66a4"/>  <img height="200" src="https://github.com/devemux86/cruiser/assets/121806238/0c6aa19d-ebbb-408f-9d09-745501dffeb1"/>  <img height="200" src="https://github.com/devemux86/cruiser/assets/121806238/3c277acf-56f1-46f6-a2d2-326af417ede2"/>
This is the imported file: [Cruiser Day5 - ACT PT.zip](https://github.com/devemux86/cruiser/files/14359250/Cruiser.Day5.-.ACT.PT.zip)

**Reply by WolfgangZ42** [reply]
Btw: I really appreciate the manual route calculation! In case of "dirty road" tracks you have to set a lot of shaping points to produce the perfect route - and it took a lot of time waiting for recalculation for every waypoint. 
Thanks for making it possible!

**Reply by devemux86** [reply]
> Is there a plan to offer additional routing options in the future?
Currently it is limited to "normal" and "beeline"...

@bios4  we need to create new profiles combining the routing settings, like in:

- https://github.com/devemux86/cruiser/discussions/254

> Use different routing options on each route section

I mean that we can change routing options and select "Recalculate section" to only change a specific route section.
e.g. change the "Avoid motorways" and use the "Extend route (start)" to calculate a quick path to an existing route.

The changed routing options are now calculated fast and only applied to the selected route section each time.
The rest of the existing route remains unchanged like how it was before.

**Reply by devemux86** [reply]
@WolfgangZ42 I moved your report at:

- https://github.com/devemux86/cruiser/discussions/449#discussioncomment-8543698

**Comment by Ciutnik** [comment]
In fact, the calculation of sections added to the end of the route is now very fast! Thanks 👍

But if the previous section was some selected alternative, this is forgotten and the "default" alternative is activated. Moreover, once you manually force a section recalculation, selecting an alternative is no longer possible. 

And one more observation: the newly added sections do not provide any alternatives to choose from, which is what I was really counting on. Maybe it would be possible to implement a selection of alternatives for newly added sections (when extending the route from the end)?

**Reply by devemux86** [reply]
@Ciutnik thanks for the report, I will check the alternative routes.

Especially BRouter, the only router generating different alternatives in each route section.

**Reply by devemux86** [reply]
@Ciutnik thanks for the detailed tests. :+1: 

Alternative routes will be fixed in the next update.

**Reply by Ciutnik** [reply]
Thanks for the quick release of 41-b2. Now alternatives work as before, at each stage of route creation they are possible for the entire route. Also, recalculating one section does not block the possibility of choosing alternatives. 👍

I would be really happy if alternatives were possible to select for each section separately!... But we probably have to wait for this possibility until we can select a profile (routing preferences as you like) for each section separately?

**Reply by devemux86** [reply]
> I would be really happy if alternatives were possible to select for each section separately!... But we probably have to wait for this possibility until we can select a profile (routing preferences as you like) for each section separately?

This could complicate the route very much, so we need to think how it can be implemented. :slightly_smiling_face: 

Essentially all routers see the alternative routes as a collection of complete paths.
Not a combination of separate alternatives in each route section.
Imagine how many combinations will be with 2-3 alternatives and 10 waypoints, for management, storage, etc.

**Reply by Ciutnik** [reply]
> Imagine how many combinations will be with 2-3 alternatives and 10 waypoints

That's true. 10 more clicks, but in return you get the perfect route! Personally, I would be willing to pay that cost 🙂
Especially since for routes with 2-3 via points it is only 2-3 clicks more. I am now thinking, for example, of very frequent, one-day trips "back and forth". Currently, it is difficult to choose an alternative that perfectly suits us in both directions of the trip.

**Comment by devemux86** [comment]
Intelligent route calculation will allow to select and combine alternatives independently on each route section.

Here is a video with a BRouter route with 3 different alternative paths in each route section.
Select independently a different alternative path on each route section and combine them into one route.

[Intelligent alternatives](https://github.com/devemux86/cruiser/assets/3484020/772fcf80-92fe-41a5-b447-bb77502e7708)

**Comment by 0709wiwiwi** [comment]
Manual mode question.
I would still like to ask who finds this new manual mode useful ? I do now struggle with the more complex Cruiser operation which was never the case before.  For (short cycling distances) I don't see the point of using manual. I prefer a direct response to my actions. And what is quite annoying to me is the fact that if you use manual and don't reset it, it is then subsequently stored in the routing profile.  And that is very annoying and surprising the next time you restart than a new design.

**Reply by devemux86** [reply]
Manual mode is optional and hidden in advanced settings, most users will continue to work with the automatic mode.

> And what is quite annoying to me is the fact that if you use manual and don't reset it, it is then subsequently stored in the routing profile. And that is very annoying and surprising the next time you restart than a new design.

What do you mean?
We move a waypoint, do not recalculate route, restart Cruiser, the waypoint is in old position.

**Reply by WolfgangZ42** [reply]
> @0709wiwiwi ... that is very annoying and surprising the next time ...

But everything you change is stored in the profile and of course you should remember your settings. Then it wouldn't be annoying/surprising you.

The manual mode in route calculation does make sense in few special planning environments, for instance: 
- If I create a route based on a gravel track (TET or ACT) I've to set a lot of shaping points. 
- Up to now each setting initiated a whole calculation of a route with >100 waypoints (~200 km).
- Now I can set several shaping points and decide when to calculate - saves time, believe me.

For a "normal" motorcycle day tour here in Bavaria, I wouldn't use the manual mode.

**Reply by 0709wiwiwi** [reply]
> of course you should remember your settings.

Of course I (will) keep forgetting that and I think I won't be the only one.
If it happens I know how to recover but it is definitely not novice friendly. 
At least Cruiser is not remaining so very KISS  by all this extra or not ? 
The exceptionally used setting manual should better so not be saved imo.

Since I didn't understand it all properly anyway, I asked the master for an explanation.
With a ready and clear answer which I think will be useful for others too. 
See text attachment.

As the whole track is no longer recalculated as before, I wonder what speed design gain you have.
It is already sure becoming more complex, so I wonder whether this extra complexity is worth it.
Since you have apparently already compiled a real track, Wolfgang, could you possibly try this out ? 

[Intelligent and Manual mode explained.txt](https://github.com/devemux86/cruiser/files/14408351/Intelligent.and.Manual.mode.explained.txt)

Note. Since we are beta testers, we should consider all possibilities for their optimal usefulness, right?

**Reply by WolfgangZ42** [reply]
I have to admit Willi, if the intelligent mode works as you have described it, then the manual mode is obsolete. As you say, we are beta testers and will wait and see how it develops.
_Now I will be in Portugal for a week and am basically "offline"!_

**Reply by bios4** [reply]
> I would still like to ask who finds this new manual mode useful ?

I do! I often plan routes for multi-day-trips with lots of waypoints, so I really like the possibility to OPTIONALLY(!) use manual mode and shuffle around multiple waypoints before ANY route (re-)calculation takes place...

> For (short cycling distances) I don't see the point of using manual.

You are right that the option makes more sense for long routes with lots of waypoints.
But then - Emux implemented the new menu button where you could switch between manual and intelligent mode by one single click.

Finally remembering the settings is clearly the task of the user... 😁 So saying an option is not useful because one is not able (or willing) to remember how he/she set it before closing the app is ... uhm ... "weird".

> [...] if the intelligent mode works as you have described it, then the manual mode is obsolete. 

I 100% disagree - both options are much appreciated and make perfectly sense depending on the routes one is working with. And as the modes can be switched easily, I really can not see any disadvantages.

**Reply by 0709wiwiwi** [reply]
I have enquired who finds manual valuable and why exactly.  So thanks. After all, for my shorter cycling trips, I do not yet see a benefit. Anyway does Cruiser still think to optimise this app also for non-motorcyclists or becomes it just an app that relies on external power feed? The new cycle season is coming soon while I still can't recommend Cruiser as cycling or walking app because the maximum energy savings possible between instructions is not addressed. Positive development for cyclist are those new BIG buttons of course.
Furthermore, I do ask that when closing the app, the state manual is not recorded in the saved profile.  If one does not like this suggestion for any reason than I would appreciate a reply.

> one is not able (or willing) to remember how he/she set it before closing the app is ... uhm ... "weird".

So it seems that apparently the perfect man (motocyclist) does exist after all.

**Reply by devemux86** [reply]
Development in all applications depends on the same things:

- **Finance**: more subscriptions can finance new features

- **Simple**: some features are easier to implement first
(especially as new Android has more restrictions)

- **Popular**: features requested by most users

Cruiser does not have a company with many developers, like other popular and "complicated" apps.
It is still a one man show all these years. :slightly_smiling_face:

**Reply by bios4** [reply]
> Furthermore, I do ask that when closing the app, the state manual is not recorded in the saved profile. If one does not like this suggestion for any reason than I would appreciate a reply.

You asked - I don't like the suggestion!
It contradicts the principle of saving ALL settings in the profile.

Instead of not saving a single setting in the profile, my suggestion would be to simply create two profiles - one with the manual option, a second one with the intelligent option.

Anyway - I don't understand why one should do this at all if you can switch the options with ONE click...

**Reply by 0709wiwiwi** [reply]
@ bios 4 
Sir, of course you can avoid this. Disable Extended Settings.
I have been active with that "other" app for a long time. I have than suggested features (knowledgeable at the time) that you only need exceptionally, yet are very detrimental in "normal" use. Especially don't underestimate the appeal of "special settings". Unbelievable but many novices immediately find and switch on extra settings without knowing why you need them and are then surprised by bad experiences. One then sometimes all too quickly uninstalls the app followed by bad app reviews. Others after years of bad experience still go and ask for advice in the forum where they then get correct advice. "If you don't know why extra is needed you shouldn't activate it either ".  A beta tester doesn't just look ahead but should also take a look in the mirror. When testing, a beta tester should also check how new users will experience the app. Before they understand the basic app, they find that extra button all too quickly and promote themselves to more experienced users. So they start using extras in what they can't handle.

**Reply by bios4** [reply]
And to make it easier for new users or beginners you want to remove (great) settings/options for all, including experienced and power users? Is this what you're suggesting?

If yes I have to admit that I don't see it like that.

And I am far from being a "Sir", by the way... 😉

**Reply by 0709wiwiwi** [reply]
okido bios4 biker.

- I am also not asking to remove the Manual option entirely.
- I ask to not save **(only) Manual** in the active profile on exit.
- Through Intelligent mode, I see new nice long-distance capabilities. 
- If there are unexpected (careless) manual remnants in the profile it fails.
- My idea is still maturing and I haven't finished this as a text at all.

I will try to publish later why this Intelligent mode could **possibly** be extra useful.
Unless I myself write off this still maturing idea as useless. He who does not try does not win.

**Reply by bios4** [reply]
> I ask to not save (only) Manual in the active profile on exit.

Explain this to all users who proactively select the manual option and wonder why it is NOT stored in the profile then.
I think this might lead to even bigger requests/questions...

**Reply by 0709wiwiwi** [reply]
I did put a fast text note together see in attachment.  Please use translate.  
I put to much time in multiple gps forums, have also other things to do sometimes.

[Intelligent to recalculate.txt](https://github.com/devemux86/cruiser/files/14421854/Intelligent.to.recalculate.txt)

**Reply by devemux86** [reply]
@0709wiwiwi I replied to this feature's topic:
- https://github.com/devemux86/cruiser/discussions/361#discussioncomment-8607011

**Comment by devemux86** [comment, marked as answer]
Intelligent route calculation of route sections is available in [Cruiser 3.0.41](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Comment by devemux86** [comment]
**Intelligent calculation of route sections** will be available with a _new implementation_ in the next Cruiser.

It will be more flexible and can work with the advanced features of BRouter (offline):
(at least in the sections without change in the route profile, e.g. fastest to curvy)

- Avoid u-turns

- Correct misplaced waypoints

- Use straight line at distant waypoints

**Comment by devemux86** [comment]
- **Intelligent calculation of route sections** is available with a new implementation in https://github.com/devemux86/cruiser/discussions/776
